File: sdk/javascript/src/uri.js
/**
 * FoodBlock URI scheme: fb:<hash> or fb:<type>/<name>
 *
 * Usage:
 *   fb.toURI(block) => 'fb:a1b2c3...'
 *   fb.toURI(block, { alias: 'sourdough' }) => 'fb:substance.product/sourdough'
 *   fb.fromURI('fb:a1b2c3...') => { hash: 'a1b2c3...' }
 *   fb.fromURI('fb:substance.product/sourdough') => { type: 'substance.product', alias: 'sourdough' }
 */

const URI_PREFIX = 'fb:'

/**
 * Convert a block or hash to a FoodBlock URI.
 * @param {object|string} blockOrHash - A block object or hash string
 * @param {object} [opts] - { alias: string } for named URIs
 * @returns {string} FoodBlock URI
 */
function toURI(blockOrHash, opts = {}) {
  if (opts.alias) {
    const type = typeof blockOrHash === 'string' ? null : blockOrHash.type
    if (type) {
      return `${URI_PREFIX}${type}/${opts.alias}`
    }
  }
  const hash = typeof blockOrHash === 'string' ? blockOrHash : blockOrHash.hash
  return `${URI_PREFIX}${hash}`
}

/**
 * Parse a FoodBlock URI.
 * @param {string} uri - A fb: URI string
 * @returns {object} { hash } for hash URIs, { type, alias } for named URIs
 */
function fromURI(uri) {
  if (!uri.startsWith(URI_PREFIX)) {
    throw new Error(`FoodBlock: invalid URI, must start with "${URI_PREFIX}"`)
  }
  const body = uri.slice(URI_PREFIX.length)

  // Check if it's a named URI: type/alias
  const slashIdx = body.indexOf('/')
  if (slashIdx !== -1 && body.indexOf('.') < slashIdx) {
    return {
      type: body.slice(0, slashIdx),
      alias: body.slice(slashIdx + 1)
    }
  }

  // Hash URI
  return { hash: body }
}

module.exports = { toURI, fromURI }

File: sdk/go/uri.go
package foodblock

import (
	"errors"
	"strings"
)

const uriPrefix = "fb:"

// ToURI converts a block or hash to a FoodBlock URI.
func ToURI(block *Block, alias string) string {
	if alias != "" && block != nil && block.Type != "" {
		return uriPrefix + block.Type + "/" + alias
	}
	if block != nil {
		return uriPrefix + block.Hash
	}
	return uriPrefix
}

// ToURIFromHash converts a hash string to a FoodBlock URI.
func ToURIFromHash(hash string) string {
	return uriPrefix + hash
}

// URIResult holds the parsed result of a FoodBlock URI.
type URIResult struct {
	Hash  string
	Type  string
	Alias string
}

// FromURI parses a FoodBlock URI.
func FromURI(uri string) (URIResult, error) {
	if !strings.HasPrefix(uri, uriPrefix) {
		return URIResult{}, errors.New("FoodBlock: invalid URI, must start with \"" + uriPrefix + "\"")
	}
	body := uri[len(uriPrefix):]

	slashIdx := strings.Index(body, "/")
	dotIdx := strings.Index(body, ".")
	if slashIdx != -1 && dotIdx != -1 && dotIdx < slashIdx {
		return URIResult{
			Type:  body[:slashIdx],
			Alias: body[slashIdx+1:],
		}, nil
	}

	return URIResult{Hash: body}, nil
}
